src/session_pool/session_pool.js

```diff
@@ -0,0 +1,148 @@
+// TODO: Validation
+export default class SessionPool extends EventEmitter {
+    /**
+     *
+     * @param [options]
+     * @param options.initialPoolSize {Number}
+     * @param options.maxPoolSize {Number}
+     * @param options.maxSessionAgeSecs {Number}
+     * @param options.maxSessionReuseCount {Number}
+     * @param options.persistStateKeyValueStoreId {String}
+     */
+    constructor(options) {
+        super();
+
+        const opts = _.defaults({}, options, SESSION_POOL_DEFAULTS);
+
+        // Pool Configuration
+        this.maxPoolSize = opts.maxPoolSize;
+        this.createSessionFunction = opts.createSessionFunction || this._createSessionFunction;
+
+        // Session configuration
+        // @TODO: Maybe options.sessionOptions / this.sessionOptions?
+        this.maxSessionAgeSecs = opts.maxSessionAgeSecs;
+        this.maxSessionReuseCount = opts.maxSessionReuseCount;
+
+        // Session storage
+        this.persistStateKeyValueStoreId = opts.persistStateKeyValueStoreId;
+        this.persistStateKey = opts.persistStateKey;
+
+        // Statistics
+        this.activeSessions = 0;
+        this.blockedSessions = 0;
+
+        // Operative states
+        this.storage = null;
+        this.sessions = [];
+
+        // Maybe we can add onSessionRetired function to configuration ?
+    }
+
+
+    async initialize() {
+        this.storage = await openKeyValueStore(this.persistStateKeyValueStoreId);
+
+        // Load sessions from storage
+        const loadedSessions = this.storage.getValue(this.persistStateKey) || [];
+
+        // Invalidate old sessions and load active sessions only
+        for (const [sessionName, sessionObject] of Object.entries(loadedSessions)) {
+            const session = Session.recreateSession(sessionObject);
+            const isValid = !session.isExpired() || !session.isBlocked() || session.usedCount >= session.maxReuseCount;
+
+            if (isValid) {
+                this.sessions[sessionName] = session;
+            }
+        }
+    }
+
+    async retrieveSession() {
+        // If we have enough space for session. Return newly created session.
+        if (this.sessions.length < this.maxPoolSize) {
+            const newSession = await this.createSessionFunction(this);
+            this._addSession(newSession);
+            return newSession;
+        }
+
+
+        // For example that developer can plug different picking algorithms such as the Lukášův
+        // Maybe a should have pickSession function to be customizable.
+        const pickedSession = this.sessions[this._getRandomIndex()];
+
+        // If session can be used return the session
+        if (pickedSession.isUsable()) {
+            return pickedSession;
+        }
+
+        //  otherwise remove old session and return newly created session
+        this._removeSession(pickedSession);
+
+        const newSession = this.createSessionFunction(this);
+        this._addSession(newSession);
+        return newSession;
```

Comment:
There are three different places where a session can be returned in this function, could we refactor? I like "returning fast", but I think it only makes sense in case of errors, missing params, etc.

Reply:
Also, the last 3 lines are the same as the lines in the first `if`